File: 1539-diagonal-traverse-ii/1539-diagonal-traverse-ii.java
class Solution {
    public int[] findDiagonalOrder(List<List<Integer>> nums) {
        int n = nums.size();
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> {
            if (a[0] == b[0])
                return b[1] - a[1];
            return a[0] - b[0];
        });

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < nums.get(i).size(); j++) {
                int sum = i + j;
                pq.offer(new int[] { sum, i, j });
            }
        }

        List<Integer> ans = new ArrayList<>();

        while (!pq.isEmpty()) {
            int r = pq.peek()[1];
            int c = pq.peek()[2];
            pq.poll();
            ans.add(nums.get(r).get(c));
        }

        int size = ans.size();
        var res = new int[size];
        for (int i = 0; i < size; i++)
            res[i] = ans.get(i);

        return res;
    }
}

File: 1539-diagonal-traverse-ii/1539-diagonal-traverse-ii.cpp
class Solution {
private:
    struct cmp {
        bool operator()(const tuple<int, int, int>& a, const tuple<int, int, int>& b) {
            if (get<0>(a) == get<0>(b))
                return get<1>(b) > get<1>(a);
            return get<0>(a) > get<0>(b);
        }
    };

public:
    vector<int> findDiagonalOrder(vector<vector<int>>& nums) {
        int n = nums.size();
        priority_queue<tuple<int, int, int>, vector<tuple<int, int, int>>, cmp> pq;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < nums[i].size(); j++) {
                int sum = i + j;
                pq.push(make_tuple(sum, i, j));
            }
        }
        
        vector<int> ans;
        while (!pq.empty()) {
            auto t = pq.top();
            auto r = get<1>(t);
            auto c = get<2>(t);
            pq.pop();

            ans.push_back(nums[r][c]);
        }

        return ans;
    }
};